README.md
# NetApp Solidfire Exporter

![Go](https://github.com/mjavier2k/solidfire-exporter/workflows/Go/badge.svg?event=push)
[![Go Report Card](https://goreportcard.com/badge/mjavier2k/solidfire-exporter)](https://goreportcard.com/report/mjavier2k/solidfire-exporter)

NetApp Solidfire-Exporter queries the Solidfire API and exports the results as Prometheus metrics


![Cluster Metrics](dashboards/solidfire-cluster-overview.jpg?raw=true)

<!--  Table of Contents Generated by the VS Code "markdown all in one" extension: yzhang.markdown-all-in-one -->
- [NetApp Solidfire Exporter](#netapp-solidfire-exporter)
  - [Supported ElementOS Versions](#supported-elementos-versions)
  - [Installation](#installation)
  - [Metrics](#metrics)
  - [Usage](#usage)
  - [Configuration](#configuration)
      - [Shell-Like Environments](#shell-like-environments)
      - [Docker-Type Environments / SystemD EnvironmentFile Environment](#docker-type-environments--systemd-environmentfile-environment)
  - [Prometheus Configuration](#prometheus-configuration)
  - [Using Docker](#using-docker)
  - [Grafana Dashboards](#grafana-dashboards)
  - [Contributing](#contributing)
  - [License](#license)

## Supported ElementOS Versions

| API Version | Endpoint | Notes |
|-|-|-|
| [11.3](https://library.netapp.com/ecm/ecm_download_file/ECMLP2856155) | https://your-mgmt-vip/json-rpc/11.3 | There is a known bug in v11.3 where the api user requires Administrator access to retrieve the QoS data. |
| [12.2](https://docs.netapp.com/sfe-122/index.jsp) | https://your-mgmt-vip/json-rpc/12.2 |  |

## Installation

Binaries can be downloaded from [Github releases](https://github.com/mjavier2k/solidfire-exporter/releases) page. 


## Metrics

| Metric | Type | Description |
|-|-|-|
| solidfire_cluster_active_block_space_bytes | gauge | The amount of space on the block drives. This includes additional information such as metadata entries and space which can be cleaned up. |
| solidfire_cluster_active_sessions | gauge | The number of active iSCSI sessions communicating with the cluster. |
| solidfire_cluster_average_io_bytes | gauge | Average size in bytes of recent I/O to the cluster in the last 500 milliseconds. |
| solidfire_cluster_average_iops | gauge | The average IOPS for the cluster since midnight Coordinated Universal Time (UTC) |
| solidfire_cluster_block_fullness | gauge | The current computed level of block fullness of the cluster. |
| solidfire_cluster_client_queue_depth | gauge | The number of outstanding read and write operations to the cluster. |
| solidfire_cluster_recent_io_size_bytes | gauge | The average size of IOPS to all volumes in the cluster |
| solidfire_cluster_compression_factor | gauge | The cluster compression factor. compressionFactor = (uniqueBlocks * 4096) / (uniqueBlocksUsedSpace * 0.93) |
| solidfire_cluster_current_iops | gauge | The average IOPS for all volumes in the cluster over the last 5 seconds |
| solidfire_cluster_de_duplication_factor | gauge | The cluster deDuplication factor. deDuplicationFactor = (nonZeroBlocks + snapshotNonZeroBlocks) / uniqueBlocks |
| solidfire_cluster_efficiency_factor | gauge | The cluster efficiency factor. efficiencyFactor = thinProvisioningFactor * deDuplicationFactor * compressionFactor |
| solidfire_cluster_fullness | gauge | Reflects the highest level of fullness between 'blockFullness' and 'metadataFullness'. |
| solidfire_cluster_iops | gauge | Current actual IOPS for the entire cluster in the last 500 milliseconds. |
| solidfire_cluster_iops_total | counter | The total number of I/O operations performed throughout the lifetime of the cluster. |
| solidfire_cluster_last_sample_read_bytes | gauge | The total number of bytes read from the cluster during the last sample period. |
| solidfire_cluster_last_sample_read_ops | gauge | The total number of read operations during the last sample period. |
| solidfire_cluster_last_sample_write_bytes | gauge | The total number of bytes written to the cluster during the last sample period. |
| solidfire_cluster_last_sample_write_ops | gauge | The total number of write operations during the last sample period. |
| solidfire_cluster_latency_seconds | gauge | The average time, in seconds, to complete operations to a cluster in the last 500 milliseconds. |
| solidfire_cluster_max_iops | gauge | The estimated maximum IOPS capability of the current cluster |
| solidfire_cluster_max_metadata_over_provision_factor | gauge | A value representative of the number of times metadata space can be over provisioned relative to the amount of space available. |
| solidfire_cluster_max_over_provisionable_space_bytes | gauge | The maximum amount of provisionable space. This is a computed value. You cannot create new volumes if the current provisioned space plus the new volume size would exceed this number. The value is calculated as follows: maxOverProvisionableSpace = maxProvisionedSpace * maxMetadataOverProvisionFactor |
| solidfire_cluster_max_provisioned_space_bytes | gauge | The total amount of provisionable space if all volumes are 100% filled (no thin provisioned metadata) |
| solidfire_cluster_max_used_metadata_space_bytes | gauge | The number of bytes on volume drives used to store metadata |
| solidfire_cluster_max_used_space_bytes | gauge | The total amount of space on all active block drives |
| solidfire_cluster_metadata_fullness | gauge | The current computed level of metadata fullness of the cluster. |
| solidfire_cluster_non_zero_blocks | gauge | The total number of 4KiB blocks that contain data after the last garbage collection operation has completed |
| solidfire_cluster_normalized_iops | gauge | Average number of IOPS for the entire cluster in the last 500 milliseconds. |
| solidfire_cluster_peak_active_sessions | gauge | The peak number of iSCSI connections since midnight UTC |
| solidfire_cluster_peak_iops | gauge | The highest value for currentIOPS since midnight UTC |
| solidfire_cluster_provisioned_space_bytes | gauge | The total amount of space provisioned in all volumes on the cluster |
| solidfire_cluster_read_bytes_total | counter | The total cumulative bytes read from the cluster since the creation of the cluster. |
| solidfire_cluster_read_latency_seconds | gauge | The average time, in seconds, to complete read operations to the cluster in the last 500 milliseconds. |
| solidfire_cluster_read_latency_seconds_total | counter | The total time spent performing read operations since the creation of the cluster. |
| solidfire_cluster_read_ops_total | counter | The total cumulative read operations to the cluster since the creation of the cluster. |
| solidfire_cluster_sample_period_seconds | gauge | The length of the sample period, in seconds. |
| solidfire_cluster_services_expected | gauge | The total number of expected services running on the cluster. |
| solidfire_cluster_services_running | gauge | The number of services running on the cluster. If equal to the servicesTotal, this indicates that valid statistics were collected from all nodes. |
| solidfire_cluster_slice_reserve_used_threshold_percentage | gauge | Error condition. A system alert is triggered if the reserved slice utilization is greater than the sliceReserveUsedThresholdPct value returned. |
| solidfire_cluster_snapshot_non_zero_blocks | gauge | The total number of 4KiB blocks that contain data after the last garbage collection operation has completed |
| solidfire_cluster_stage2_aware_threshold_percentage | gauge | Awareness condition. The value that is set for 'Stage 2' cluster threshold level. |
| solidfire_cluster_stage2_block_threshold_bytes | gauge | Number of bytes being used by the cluster at which a stage2 condition will exist. |
| solidfire_cluster_stage3_block_threshold_bytes | gauge | Number of bytes being used by the cluster at which a stage3 condition will exist. |
| solidfire_cluster_stage3_block_threshold_percentage | gauge | Percent value set for stage3. At this percent full, a warning is posted in the Alerts log. |
| solidfire_cluster_stage3_low_threshold_percentage | gauge | Error condition. The threshold at which a system alert is created due to low capacity on a cluster |
| solidfire_cluster_stage4_block_threshold_bytes | gauge | Number of bytes being used by the cluster at which a stage4 condition will exist |
| solidfire_cluster_stage4_critical_threshold_percentage | gauge | Error condition. The threshold at which a system alert is created to warn about critically low capacity on a cluster. |
| solidfire_cluster_stage5_block_threshold_bytes | gauge | The number of bytes being used by the cluster at which a stage5 condition will exist. |
| solidfire_cluster_thin_provisioning_factor | gauge | The cluster thin provisioning factor. thinProvisioningFactor = (nonZeroBlocks + zeroBlocks) / nonZeroBlocks |
| solidfire_cluster_throughput_utilization | gauge | The cluster capacity being utilized. 0 - not utilized. 1 - 100% utilized. |
| solidfire_cluster_total_bytes | gauge | Physical capacity of the cluster, measured in bytes. |
| solidfire_cluster_total_metadata_bytes | gauge | Total amount of space that can be used to store metadata |
| solidfire_cluster_unaligned_reads_total | counter | The total cumulative unaligned read operations to a cluster since the creation of the cluster |
| solidfire_cluster_unaligned_writes_total | counter | The total cumulative unaligned write operations to a cluster since the creation of the cluster. |
| solidfire_cluster_unique_blocks | gauge | The total number of blocks stored on the block drives The value includes replicated blocks |
| solidfire_cluster_unique_blocks_used_space_bytes | gauge | The total amount of data the uniqueBlocks take up on the block drives |
| solidfire_cluster_used_bytes | gauge | Number of bytes used on the cluster. |
| solidfire_cluster_used_metadata_bytes | gauge | Amount of space used on volume drives to store metadata. |
| solidfire_cluster_used_metadata_space_bytes | gauge | The total number of bytes on volume drives used to store metadata |
| solidfire_cluster_used_metadata_space_in_snapshots_bytes | gauge | The number of bytes on volume drives used for storing unique data in snapshots. This number provides an estimate of how much metadata space would be regained by deleting all snapshots on the system |
| solidfire_cluster_used_space_bytes | gauge | The total amount of space used by all block drives in the system |
| solidfire_cluster_write_bytes_total | counter | The total cumulative bytes written to the cluster since the creation of the cluster |
| solidfire_cluster_write_latency_seconds | gauge | The average time, in seconds, to complete write operations to a cluster in the last 500 milliseconds. |
| solidfire_cluster_write_latency_seconds_total | counter | The total time spent performing write operations since the creation of the cluster. |
| solidfire_cluster_write_ops_total | counter | The total cumulative write operations to the cluster since the creation of the cluster. |
| solidfire_cluster_zero_blocks | gauge | The total number of empty 4KiB blocks without data after the last round of garbage collection operation has completed |
| solidfire_drive_capacity_bytes | gauge | The drive capacity for each individual drives in the cluster's active nodes |
| solidfire_drive_status | gauge | The drive status for each individual drives in the cluster's active nodes |

CHANGELOG.md
# Changelog
<!-- https://keepachangelog.com/en/1.0.0/

Types of Changes:
### Added - for new features.
### Changed - for changes in existing functionality.
### Deprecated - for soon-to-be removed features.
### Removed - for now removed features.
### Fixed - for any bug fixes.
### Security - in case of vulnerabilities.

Example Format Below:

## [0.0.7] - 2015-02-16
### Added
- Link, and make it obvious that date format is ISO 8601.

### Changed
- Clarified the section on "Is there a standard change log format?".

### Fixed
- Fix Markdown links to tag comparison URL with footnote-style links.
-->
## [Unreleased]
## [0.6.2] - 2021-07-30
### Fixed
- avoid panic when reading maps #68
## [0.6.1] - 2021-05-25
### Added
- Call the SF API concurrently for scrape speed improvement #62
- Added global scrape timeout #64
- Added tests for collector #60

### Fixed
- Sum ISCSI sessions for all volumes per node in singlestat panel #59 

## [0.6.0] - 2021-04-17
### Added
- Added CHANGELOG
- Added new versions of cluster overview, node detail, and volume detail dashboard powered by grabana. These new dashboards use the new metric names.
### Changed
- BREAKING CHANGE: Most metric names have been updated to follow Prometheus conventions. See README for the new metric names and types.

### Deprecated
- Marked old json dashboards as deprecated in examples/dashboards/README.md

## [0.5.2] - 2021-03-11

### Added
- Added license [#32](https://github.com/mjavier2k/solidfire-exporter/pull/32)
- Added official dashboards [#39](https://github.com/mjavier2k/solidfire-exporter/pull/39)
- Added Arm64 Artifact [#46](https://github.com/mjavier2k/solidfire-exporter/pull/46)
- Officially supports ElementOS 12 [#48](https://github.com/mjavier2k/solidfire-exporter/pull/48)

### Changes
- Changed scrape_success metric to solidfire_up [#29](https://github.com/mjavier2k/solidfire-exporter/pull/29)
- Added option to specify config.yaml (NOTE: Environment variable names changed) [#30](https://github.com/mjavier2k/solidfire-exporter/pull/30)

### Fixed
- Protect against bad URI causing crash [#47](https://github.com/mjavier2k/solidfire-exporter/pull/47)
- Fix efficiency factor calculation [#49](https://github.com/mjavier2k/solidfire-exporter/pull/49)


pkg/solidfire/types.go
package solidfire

import (
	"net/http"
	"time"
)

const (
	ListenAddress        string = "listen.address"
	DefaultListenAddress string = "0.0.0.0:9987"

	Username        string = "client.username"
	DefaultUsername string = "default"
	Password        string = "client.password"
	DefaultPassword string = "default"

	Endpoint        string = "client.endpoint"
	DefaultEndpoint string = "http://127.0.0.1/json-rpc/11.3"
	InsecureSSL     string = "client.insecure"

	HTTPClientTimeout        string = "client.timeout"
	DefaultHTTPClientTimeout int    = 30

	LogLevel        string = "log.level"
	DefaultLogLevel        = "info"

	ConfigFile        string = "config"
	DefaultConfigFile string = "config.yaml"
)

type Client struct {
	Username    string
	Password    string
	RPCEndpoint string
	HttpClient  *http.Client
}
type Interface interface {
	GetClusterCapacity() (GetClusterCapacityResponse, error)
	GetClusterFullThreshold() (GetClusterFullThresholdResponse, error)
	GetClusterStats() (GetClusterStatsResponse, error)
	ListAllNodes() (ListAllNodesResponse, error)
	ListClusterFaults() (ListClusterFaultsResponse, error)
	ListDrives() (ListDrivesResponse, error)
	ListISCSISessions() (ListISCSISessionsResponse, error)
	ListNodeStats() (ListNodeStatsResponse, error)
	ListVolumeQoSHistograms() (ListVolumeQoSHistogramsResponse, error)
	ListVolumes() (ListVolumesResponse, error)
	ListVolumeStats() (ListVolumeStatsResponse, error)
}
type RPCBody struct {
	Method RPC       `json:"method"`
	Params RPCParams `json:"params"`
	ID     int       `json:"id"`
}
type RPCParams interface{}
type ListVolumesRPCParams struct {
	IncludeVirtualVolumes bool `json:"includeVirtualVolumes"`
}
type ListVolumeStatsRPCParams struct {
	VolumeIDs             []int `json:"volumeIDs"`
	IncludeVirtualVolumes bool  `json:"includeVirtualVolumes"`
}
type GetClusterCapacityRPCParams struct {
	// No params needed
}

type ListClusterFaultsRPCParams struct {
	FaultTypes    string `json:"faultTypes"`
	BestPractices bool   `json:"bestPractices"`
}

type ListNodeStatsRPCParams struct {
	// No params needed
}

type ListVolumeQoSHistogramsRPCParams struct {
	VolumeIDs []int `json:"volumeIDs"`
}

type ListAllNodesRPCParams struct {
	// No params needed
}
type GetClusterStatsRPCParams struct {
	// No params needed
}

type GetClusterFullThresholdParams struct {
	// No params needed
}

type ListDrivesParams struct {
	// No params needed
}

type ListISCSISessionsParams struct {
	// No params needed
}

type ListVolumesResponse struct {
	ID     int `json:"id"`
	Result struct {
		Volumes []struct {
			Access     string `json:"access"`
			AccountID  int    `json:"accountID"`
			Attributes struct {
			} `json:"attributes"`
			BlockSize                   int         `json:"blockSize"`
			CreateTime                  time.Time   `json:"createTime"`
			CurrentProtectionScheme     string      `json:"currentProtectionScheme"`
			DeleteTime                  string      `json:"deleteTime"`
			Enable512E                  bool        `json:"enable512e"`
			EnableSnapMirrorReplication bool        `json:"enableSnapMirrorReplication"`
			Iqn                         string      `json:"iqn"`
			LastAccessTime              time.Time   `json:"lastAccessTime"`
			LastAccessTimeIO            time.Time   `json:"lastAccessTimeIO"`
			Name                        string      `json:"name"`
			PreviousProtectionScheme    interface{} `json:"previousProtectionScheme"`
			PurgeTime                   string      `json:"purgeTime"`
			Qos                         struct {
				BurstIOPS int `json:"burstIOPS"`
				BurstTime int `json:"burstTime"`
				Curve     struct {
					Num4096    int `json:"4096"`
					Num8192    int `json:"8192"`
					Num16384   int `json:"16384"`
					Num32768   int `json:"32768"`
					Num65536   int `json:"65536"`
					Num131072  int `json:"131072"`
					Num262144  int `json:"262144"`
					Num524288  int `json:"524288"`
					Num1048576 int `json:"1048576"`
				} `json:"curve"`
				MaxIOPS int `json:"maxIOPS"`
				MinIOPS int `json:"minIOPS"`
			} `json:"qos"`
			QosPolicyID                interface{}   `json:"qosPolicyID"`
			ScsiEUIDeviceID            string        `json:"scsiEUIDeviceID"`
			ScsiNAADeviceID            string        `json:"scsiNAADeviceID"`
			SliceCount                 int           `json:"sliceCount"`
			Status                     string        `json:"status"`
			TotalSize                  int64         `json:"totalSize"`
			VirtualVolumeID            interface{}   `json:"virtualVolumeID"`
			VolumeAccessGroups         []interface{} `json:"volumeAccessGroups"`
			VolumeConsistencyGroupUUID string        `json:"volumeConsistencyGroupUUID"`
			VolumeID                   int           `json:"volumeID"`
			VolumePairs                []interface{} `json:"volumePairs"`
			VolumeUUID                 string        `json:"volumeUUID"`
		} `json:"volumes"`
	} `json:"result"`
}

type ListVolumeStatsResponse struct {
	ID     int `json:"id"`
	Result struct {
		VolumeStats []struct {
			AccountID            int64       `json:"accountID"`
			ActualIOPS           float64     `json:"actualIOPS"`
			AsyncDelay           interface{} `json:"asyncDelay"`
			AverageIOPSize       float64     `json:"averageIOPSize"`
			BurstIOPSCredit      float64     `json:"burstIOPSCredit"`
			ClientQueueDepth     float64     `json:"clientQueueDepth"`
			DesiredMetadataHosts interface{} `json:"desiredMetadataHosts"`
			LatencyUSec          float64     `json:"latencyUSec"`
			// MetadataHosts        struct {
			// 	DeadSecondaries []interface{} `json:"deadSecondaries"`
			// 	LiveSecondaries []int         `json:"liveSecondaries"`
			// 	Primary         int           `json:"primary"`
			// } `json:"metadataHosts"`
			NonZeroBlocks         float64       `json:"nonZeroBlocks"`
			NormalizedIOPS        float64       `json:"normalizedIOPS"`
			ReadBytes             float64       `json:"readBytes"`
			ReadBytesLastSample   float64       `json:"readBytesLastSample"`
			ReadLatencyUSec       float64       `json:"readLatencyUSec"`
			ReadLatencyUSecTotal  float64       `json:"readLatencyUSecTotal"`
			ReadOps               float64       `json:"readOps"`
			ReadOpsLastSample     float64       `json:"readOpsLastSample"`
			SamplePeriodMSec      float64       `json:"samplePeriodMSec"`
			Throttle              float64       `json:"throttle"`
			Timestamp             time.Time     `json:"timestamp"`
			UnalignedReads        float64       `json:"unalignedReads"`
			UnalignedWrites       float64       `json:"unalignedWrites"`
			VolumeAccessGroups    []interface{} `json:"volumeAccessGroups"`
			VolumeID              int           `json:"volumeID"`
			VolumeSize            float64       `json:"volumeSize"`
			VolumeUtilization     float64       `json:"volumeUtilization"`
			WriteBytes            float64       `json:"writeBytes"`
			WriteBytesLastSample  float64       `json:"writeBytesLastSample"`
			WriteLatencyUSec      float64       `json:"writeLatencyUSec"`
			WriteLatencyUSecTotal float64       `json:"writeLatencyUSecTotal"`
			WriteOps              float64       `json:"writeOps"`
			WriteOpsLastSample    float64       `json:"writeOpsLastSample"`
			ZeroBlocks            float64       `json:"zeroBlocks"`
		} `json:"volumeStats"`
	} `json:"result"`
}
type GetClusterCapacityResponse struct {
	ID     int `json:"id"`
	Result struct {
		ClusterCapacity struct {
			ActiveBlockSpace             float64   `json:"activeBlockSpace"`
			ActiveSessions               float64   `json:"activeSessions"`
			AverageIOPS                  float64   `json:"averageIOPS"`
			ClusterRecentIOSize          float64   `json:"clusterRecentIOSize"`
			CurrentIOPS                  float64   `json:"currentIOPS"`
			MaxIOPS                      float64   `json:"maxIOPS"`
			MaxOverProvisionableSpace    float64   `json:"maxOverProvisionableSpace"`
			MaxProvisionedSpace          float64   `json:"maxProvisionedSpace"`
			MaxUsedMetadataSpace         float64   `json:"maxUsedMetadataSpace"`
			MaxUsedSpace                 float64   `json:"maxUsedSpace"`
			NonZeroBlocks                float64   `json:"nonZeroBlocks"`
			PeakActiveSessions           float64   `json:"peakActiveSessions"`
			PeakIOPS                     float64   `json:"peakIOPS"`
			ProvisionedSpace             float64   `json:"provisionedSpace"`
			SnapshotNonZeroBlocks        float64   `json:"snapshotNonZeroBlocks"`
			Timestamp                    time.Time `json:"timestamp"`
			TotalOps                     float64   `json:"totalOps"`
			UniqueBlocks                 float64   `json:"uniqueBlocks"`
			UniqueBlocksUsedSpace        float64   `json:"uniqueBlocksUsedSpace"`
			UsedMetadataSpace            float64   `json:"usedMetadataSpace"`
			UsedMetadataSpaceInSnapshots float64   `json:"usedMetadataSpaceInSnapshots"`
			UsedSpace                    float64   `json:"usedSpace"`
			ZeroBlocks                   float64   `json:"zeroBlocks"`
		} `json:"clusterCapacity"`
	} `json:"result"`
}
type ListClusterFaultsResponse struct {
	ID     int `json:"id"`
	Result struct {
		Faults []struct {
			BlocksUpgrade       bool          `json:"blocksUpgrade"`
			ClusterFaultID      float64       `json:"clusterFaultID"`
			Code                string        `json:"code"`
			Data                interface{}   `json:"data"`
			Date                interface{}   `json:"date"`
			Details             string        `json:"details"`
			DriveID             float64       `json:"driveID"`
			DriveIDs            []interface{} `json:"driveIDs"`
			ExternalSource      string        `json:"externalSource"`
			NetworkInterface    string        `json:"networkInterface"`
			NodeHardwareFaultID float64       `json:"nodeHardwareFaultID"`
			NodeID              int           `json:"nodeID"`
			Resolved            bool          `json:"resolved"`
			ResolvedDate        interface{}   `json:"resolvedDate"`
			ServiceID           float64       `json:"serviceID"`
			Severity            string        `json:"severity"`
			Type                string        `json:"type"`
		} `json:"faults"`
	} `json:"result"`
}

type ListNodeStatsResponse struct {
	ID     int `json:"id"`
	Result struct {
		NodeStats struct {
			Nodes []struct {
				CBytesIn                  float64 `json:"cBytesIn"`
				CBytesOut                 float64 `json:"cBytesOut"`
				Count                     uint64  `json:"count"`
				CPU                       float64 `json:"cpu"`
				CPUTotal                  float64 `json:"cpuTotal"`
				MBytesIn                  float64 `json:"mBytesIn"`
				MBytesOut                 float64 `json:"mBytesOut"`
				NetworkUtilizationCluster float64 `json:"networkUtilizationCluster"`
				NetworkUtilizationStorage float64 `json:"networkUtilizationStorage"`
				NodeID                    int     `json:"nodeID"`
				ReadLatencyUSecTotal      float64 `json:"readLatencyUSecTotal"`
				ReadOps                   float64 `json:"readOps"`
				SBytesIn                  float64 `json:"sBytesIn"`
				SBytesOut                 float64 `json:"sBytesOut"`
				SsLoadHistogram           struct {
					Bucket0       uint64 `json:"Bucket0"`
					Bucket1To19   uint64 `json:"Bucket1To19"`
					Bucket20To39  uint64 `json:"Bucket20To39"`
					Bucket40To59  uint64 `json:"Bucket40To59"`
					Bucket60To79  uint64 `json:"Bucket60To79"`
					Bucket80To100 uint64 `json:"Bucket80To100"`
				} `json:"ssLoadHistogram"`
				Timestamp             time.Time `json:"timestamp"`
				UsedMemory            float64   `json:"usedMemory"`
				WriteLatencyUSecTotal float64   `json:"writeLatencyUSecTotal"`
				WriteOps              float64   `json:"writeOps"`
			} `json:"nodes"`
		} `json:"nodeStats"`
	} `json:"result"`
}

type ListVolumeQoSHistogramsResponse struct {
	ID     int `json:"id"`
	Result struct {
		QosHistograms []struct {
			Histograms struct {
				BelowMinIopsPercentages struct {
					Bucket1To19   uint64 `json:"Bucket1To19"`
					Bucket20To39  uint64 `json:"Bucket20To39"`
					Bucket40To59  uint64 `json:"Bucket40To59"`
					Bucket60To79  uint64 `json:"Bucket60To79"`
					Bucket80To100 uint64 `json:"Bucket80To100"`
				} `json:"belowMinIopsPercentages"`
				MinToMaxIopsPercentages struct {
					Bucket101Plus uint64 `json:"Bucket101Plus"`
					Bucket1To19   uint64 `json:"Bucket1To19"`
					Bucket20To39  uint64 `json:"Bucket20To39"`
					Bucket40To59  uint64 `json:"Bucket40To59"`
					Bucket60To79  uint64 `json:"Bucket60To79"`
					Bucket80To100 uint64 `json:"Bucket80To100"`
				} `json:"minToMaxIopsPercentages"`
				ReadBlockSizes struct {
					Bucket131072Plus    uint64 `json:"Bucket131072Plus"`
					Bucket16384To32767  uint64 `json:"Bucket16384To32767"`
					Bucket32768To65535  uint64 `json:"Bucket32768To65535"`
					Bucket4096To8191    uint64 `json:"Bucket4096To8191"`
					Bucket65536To131071 uint64 `json:"Bucket65536To131071"`
					Bucket8192To16383   uint64 `json:"Bucket8192To16383"`
				} `json:"readBlockSizes"`
				TargetUtilizationPercentages struct {
					Bucket0       uint64 `json:"Bucket0"`
					Bucket101Plus uint64 `json:"Bucket101Plus"`
					Bucket1To19   uint64 `json:"Bucket1To19"`
					Bucket20To39  uint64 `json:"Bucket20To39"`
					Bucket40To59  uint64 `json:"Bucket40To59"`
					Bucket60To79  uint64 `json:"Bucket60To79"`
					Bucket80To100 uint64 `json:"Bucket80To100"`
				} `json:"targetUtilizationPercentages"`
				ThrottlePercentages struct {
					Bucket0       uint64 `json:"Bucket0"`
					Bucket1To19   uint64 `json:"Bucket1To19"`
					Bucket20To39  uint64 `json:"Bucket20To39"`
					Bucket40To59  uint64 `json:"Bucket40To59"`
					Bucket60To79  uint64 `json:"Bucket60To79"`
					Bucket80To100 uint64 `json:"Bucket80To100"`
				} `json:"throttlePercentages"`
				WriteBlockSizes struct {
					Bucket131072Plus    uint64 `json:"Bucket131072Plus"`
					Bucket16384To32767  uint64 `json:"Bucket16384To32767"`
					Bucket32768To65535  uint64 `json:"Bucket32768To65535"`
					Bucket4096To8191    uint64 `json:"Bucket4096To8191"`
					Bucket65536To131071 uint64 `json:"Bucket65536To131071"`
					Bucket8192To16383   uint64 `json:"Bucket8192To16383"`
				} `json:"writeBlockSizes"`
			} `json:"histograms"`
			Timestamp time.Time `json:"timestamp"`
			VolumeID  int       `json:"volumeID"`
		} `json:"qosHistograms"`
	} `json:"result"`
}
type ListAllNodesResponse struct {
	ID     int `json:"id"`
	Result struct {
		Nodes []struct {
			AssociatedFServiceID      int `json:"associatedFServiceID"`
			AssociatedMasterServiceID int `json:"associatedMasterServiceID"`
			Attributes                struct {
			} `json:"attributes"`
			ChassisName                 string      `json:"chassisName"`
			Cip                         string      `json:"cip"`
			Cipi                        string      `json:"cipi"`
			FibreChannelTargetPortGroup interface{} `json:"fibreChannelTargetPortGroup"`
			Mip                         string      `json:"mip"`
			Mipi                        string      `json:"mipi"`
			Name                        string      `json:"name"`
			NodeID                      int         `json:"nodeID"`
			NodeSlot                    string      `json:"nodeSlot"`
			PlatformInfo                struct {
				ChassisType           string  `json:"chassisType"`
				CPUModel              string  `json:"cpuModel"`
				NodeMemoryGB          float64 `json:"nodeMemoryGB"`
				NodeType              string  `json:"nodeType"`
				PlatformConfigVersion string  `json:"platformConfigVersion"`
			} `json:"platformInfo"`
			Sip             string        `json:"sip"`
			Sipi            string        `json:"sipi"`
			SoftwareVersion string        `json:"softwareVersion"`
			UUID            string        `json:"uuid"`
			VirtualNetworks []interface{} `json:"virtualNetworks"`
		} `json:"nodes"`
		PendingActiveNodes []interface{} `json:"pendingActiveNodes"`
		PendingNodes       []interface{} `json:"pendingNodes"`
	} `json:"result"`
}
type GetClusterStatsResponse struct {
	ID     int `json:"id"`
	Result struct {
		ClusterStats struct {
			ActualIOPS           float64 `json:"actualIOPS"`
			AverageIOPSize       float64 `json:"averageIOPSize"`
			ClientQueueDepth     float64 `json:"clientQueueDepth"`
			ClusterUtilization   float64 `json:"clusterUtilization"`
			LatencyUSec          float64 `json:"latencyUSec"`
			NormalizedIOPS       float64 `json:"normalizedIOPS"`
			ReadBytes            float64 `json:"readBytes"`
			ReadBytesLastSample  float64 `json:"readBytesLastSample"`
			ReadLatencyUSec      float64 `json:"readLatencyUSec"`
			ReadLatencyUSecTotal float64 `json:"readLatencyUSecTotal"`
			ReadOps              float64 `json:"readOps"`
			ReadOpsLastSample    float64 `json:"readOpsLastSample"`
			SamplePeriodMsec     float64 `json:"samplePeriodMsec"`
			ServicesCount        float64 `json:"servicesCount"`
			ServicesTotal        float64 `json:"servicesTotal"`
			// Timestamp             time.Time `json:"timestamp"`
			UnalignedReads        float64 `json:"unalignedReads"`
			UnalignedWrites       float64 `json:"unalignedWrites"`
			WriteBytes            float64 `json:"writeBytes"`
			WriteBytesLastSample  float64 `json:"writeBytesLastSample"`
			WriteLatencyUSec      float64 `json:"writeLatencyUSec"`
			WriteLatencyUSecTotal float64 `json:"writeLatencyUSecTotal"`
			WriteOps              float64 `json:"writeOps"`
			WriteOpsLastSample    float64 `json:"writeOpsLastSample"`
		} `json:"clusterStats"`
	} `json:"result"`
}

type GetClusterFullThresholdResponse struct {
	ID     int `json:"id"`
	Result struct {
		BlockFullness                  string  `json:"blockFullness"`
		Fullness                       string  `json:"fullness"`
		MaxMetadataOverProvisionFactor float64 `json:"maxMetadataOverProvisionFactor"`
		MetadataFullness               string  `json:"metadataFullness"`
		SliceReserveUsedThresholdPct   float64 `json:"sliceReserveUsedThresholdPct"`
		Stage2AwareThreshold           float64 `json:"stage2AwareThreshold"`
		Stage2BlockThresholdBytes      float64 `json:"stage2BlockThresholdBytes"`
		Stage3BlockThresholdBytes      float64 `json:"stage3BlockThresholdBytes"`
		Stage3BlockThresholdPercent    float64 `json:"stage3BlockThresholdPercent"`
		Stage3LowThreshold             float64 `json:"stage3LowThreshold"`
		Stage4BlockThresholdBytes      float64 `json:"stage4BlockThresholdBytes"`
		Stage4CriticalThreshold        float64 `json:"stage4CriticalThreshold"`
		Stage5BlockThresholdBytes      float64 `json:"stage5BlockThresholdBytes"`
		SumTotalClusterBytes           float64 `json:"sumTotalClusterBytes"`
		SumTotalMetadataClusterBytes   float64 `json:"sumTotalMetadataClusterBytes"`
		SumUsedClusterBytes            float64 `json:"sumUsedClusterBytes"`
		SumUsedMetadataClusterBytes    float64 `json:"sumUsedMetadataClusterBytes"`
	} `json:"result"`
}

type ListDrivesResponse struct {
	ID     int `json:"id"`
	Result struct {
		Drives []struct {
			Attributes struct {
			} `json:"attributes"`
			Capacity float64 `json:"capacity"`
			DriveID  int     `json:"driveID"`
			NodeID   int     `json:"nodeID"`
			Serial   string  `json:"serial"`
			Slot     int     `json:"slot"`
			Status   string  `json:"status"`
			Type     string  `json:"type"`
		} `json:"drives"`
	} `json:"result"`
}

type ListISCSISessionsResponse struct {
	ID     int `json:"id"`
	Result struct {
		Sessions []struct {
			AccountID              int         `json:"accountID"`
			AccountName            string      `json:"accountName"`
			CreateTime             time.Time   `json:"createTime"`
			DriveID                int         `json:"driveID"`
			DriveIDs               []int       `json:"driveIDs"`
			Initiator              interface{} `json:"initiator"`
			InitiatorIP            string      `json:"initiatorIP"`
			InitiatorName          string      `json:"initiatorName"`
			InitiatorPortName      string      `json:"initiatorPortName"`
			InitiatorSessionID     float64     `json:"initiatorSessionID"`
			MsSinceLastIscsiPDU    int         `json:"msSinceLastIscsiPDU"`
			MsSinceLastScsiCommand int         `json:"msSinceLastScsiCommand"`
			NodeID                 int         `json:"nodeID"`
			ServiceID              int         `json:"serviceID"`
			SessionID              int64       `json:"sessionID"`
			TargetIP               string      `json:"targetIP"`
			TargetName             string      `json:"targetName"`
			TargetPortName         string      `json:"targetPortName"`
			VirtualNetworkID       int         `json:"virtualNetworkID"`
			VolumeID               int         `json:"volumeID"`
			VolumeInstance         int64       `json:"volumeInstance"`
		} `json:"sessions"`
	} `json:"result"`
}
